master/internal/storage/service.go

```diff
@@ -0,0 +1,211 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/uptrace/bun"
+	"golang.org/x/exp/maps"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/schemas"
+	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
+)
+
+// AddBackend adds storage backend information.
+// We won't persist the "save_*_best" fields on the expconf struct.
+// If the same storage backend has been persisted before then we will return the ID and
+// not insert another row.
+func AddBackend(
+	ctx context.Context, cs *expconf.CheckpointStorageConfig,
+) (model.StorageBackendID, error) {
+	cs = schemas.WithDefaults(cs)
+	if err := schemas.IsComplete(cs); err != nil {
+		return 0, fmt.Errorf("schema is not complete: %w", err)
+	}
+
+	childTableRow, unionType := expconfToStorage(cs)
+
+	var res model.StorageBackendID
+	if err := db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		if _, err := tx.NewInsert().Returning("id").
+			On("CONFLICT DO NOTHING").
+			Model(childTableRow).
+			Exec(ctx); err != nil {
+			json, jsonErr := cs.MarshalJSON()
+			if jsonErr != nil {
+				return fmt.Errorf("adding storage backend: %w: %w", jsonErr, err)
+			}
+			return fmt.Errorf("adding storage backend %s: %w", string(json), err)
+		}
+
+		// ON CONFLICT DO NOTHING returns a non zero ID only when we insert a new row.
+		// When we insert a new row also insert a new row in the parent table.
+		if childTableRow.id() != 0 {
+			unionTableRow := &storageBackendRow{}
```

Comment:
I think the style guide is ambiguous on this point since it is a ptr?

I think in this case there is just one point where we need to take the address of the pointer, but if there many what would you do?

fwiw Go/Uber has cases where they do ```v := &structType{}```
https://github.com/uber-go/zap/blob/d27427d23f81dba1f048d6034d5f286572049e1e/internal/readme/readme.go#L75

https://github.com/golang/go/blob/29cea6583ead4fa5b12a333e6a9dabe3f8a52fdd/src/sync/map.go#L105